### PR TITLE
[Efficient Metadata Operations] Reserve a partitionid for Metadatachunk during PutOperation.

### PR DESCRIPTION
**Pull request by ankagrawal**

### CODE REVIEW

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterMetrics.java [hunk 1]
```diff
@@ -215,6 +215,13 @@ public class NonBlockingRouterMetrics {
   public final Counter compositeBlobGetCount;
   public final Counter rawBlobGetCount;
 
+  // Metrics for Reserving partitions and blobid for metadata chunk.
+  public final Counter numFailedPartitionReserveAttempts;
+  public final Counter unknownReservedPartitionClassCount;
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterMetrics.java:220]: Shall we use the same name convention and call it numUnknownReservedPartitionClassCount?

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 2]
```diff
@@ -1917,6 +1921,48 @@ protected PutRequest createPutRequest() {
...
+      PartitionId selected = clusterMap.getRandomFullyWritablePartition(partitionClass, partitionIdsToExclude);
+      if (selected == null) {
+        routerMetrics.numFailedPartitionReserveAttempts.inc();
+        return null;
+      }
+      if (!partitionClass.equals(selected.getPartitionClass())) {
+        logger.warn(
+            "{}: No partitions for partitionClass='{}' found, partitionClass='{}' used instead for metadata chunk. blobProperties={}",
+            loggingContext, partitionClass, selected.getPartitionClass(), passedInBlobProperties);
+        routerMetrics.unknownReservedPartitionClassCount.inc();
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1941]: Is that "unexpectedReservedPartition" better than "unknownReservedPartition"?

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 3]
```diff
@@ -1772,6 +1773,7 @@ private class MetadataPutChunk extends PutChunk {
     private final TreeMap<Integer, Pair<StoreKey, Long>> indexToChunkIdsAndChunkSizes;
     private int intermediateChunkSize = routerConfig.routerMaxPutChunkSizeBytes;
     private Pair<? extends StoreKey, BlobProperties> firstChunkIdAndProperties = null;
+    private final PartitionId reservedPartitionId;
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1776]: We don't use the reservedPartitionId yet in the PR, right?

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterMetrics.java (hunk 1 above)
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterMetrics.java]: done.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java (hunk 3 above)
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1776]: Yes. We don't use it yet. Right now we want to collect data about
1. How many partitions are RW and PRW and whats the right threshold.
2. How many metadata chunks with reserved ids will fail or need to go cross colo.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java (hunk 2 above)
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/PutOperation.java]: changed.

**Review by JingQianCloud — APPROVED** [review]

**Review by Arun-LinkedIn — APPROVED** [review]

**Pull request merged**